src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,1519 @@
+	private function process_next_tag() {
+		/*
+		 * We're done with the document but some tags
+		 * are still open. Let's close them one at a time.
+		 */
+		if ( $this->is_closing_open_tags ) {
+			// If only the root element is open, we're done.
+			if(count($this->open_elements) <= 1)
+			{
+				return false;
+			}
+
+			// Otherwise close the next open tag on the stack
+			$this->current_token = null;
+			$this->current_token_start = strlen($this->html);
+			$this->current_token_end = strlen($this->html);
+
+			$this->pop_open_element();
+			$this->get_updated_html();
+
+			parent::next_tag(array('tag_closers' => 'visit'));
+			$this->current_token = new WP_HTML_Tag_Token($this->get_tag());
+			$this->current_token_start = $this->tag_name_starts_at - 2;
+			$this->current_token_end = $this->tag_ends_at;
+			return true;
+		}
+		
+		/**
+		 * Go to the next tag and process any text was found along the way.
+		 */
+		$text_start = $this->tag_ends_at + 1;
+		if (!parent::next_tag(array('tag_closers' => 'visit'))) {
+			$this->process_text($text_start, strlen($this->html));
+
+			$this->is_closing_open_tags = true;
+			return $this->process_next_tag();
+		}
+
+		/**
+		 * We found a tag! Let's process any text we may have found along the way.
+		 */
+		$current_tag_start = $this->tag_starts_at();
+		$this->process_text($text_start, $current_tag_start);
+
+		$this->current_token = new WP_HTML_Tag_Token($this->get_tag());
+		$this->current_token_start = $current_tag_start;
+		$this->current_token_end = $this->tag_ends_at;
+
+		if ( ! $this->is_tag_closer() ) {
+			dbg( "Found {$this->current_token->tag} tag opener" );
+			switch ( $this->current_token->tag ) {
+				case 'HTML':
+					$this->drop_current_tag_token();
+					break;
+				case 'ADDRESS':
+				case 'ARTICLE':
+				case 'ASIDE':
+				case 'BLOCKQUOTE':
+				case 'CENTER':
+				case 'DETAILS':
+				case 'DIALOG':
+				case 'DIR':
+				case 'DIV':
+				case 'DL':
+				case 'FIELDSET':
+				case 'FIGCAPTION':
+				case 'FIGURE':
+				case 'FOOTER':
+				case 'HEADER':
+				case 'HGROUP':
+				case 'MAIN':
+				case 'MENU':
+				case 'NAV':
+				case 'OL':
+				case 'P':
+				case 'SECTION':
+				case 'SUMMARY':
+				case 'UL':
+					// Ignore special rules for 'PRE' and 'LISTING'
+				case 'PRE':
+				case 'LISTING':
+					/*
+					 * If the stack of open elements has a p element in button scope,
+					 * then close a p element.
+					 */
+					if ( $this->is_element_in_button_scope( 'P' ) ) {
+						$this->close_p_element();
+					}
+					$this->insert_element( $this->current_token );
+					break;
+				// A start tag whose tag name is "h1", "h2", "h3", "h4", "h5", or "h6"
+				case 'H1':
+				case 'H2':
+				case 'H3':
+				case 'H4':
+				case 'H5':
+				case 'H6':
+					if ( $this->is_element_in_button_scope( 'P' ) ) {
+						$this->close_p_element();
+					}
+					if ( in_array( $this->current_node()->tag, array( 'H1', 'H2', 'H3', 'H4', 'H5', 'H6' ) ) ) {
```

Comment:
isn't this implied by the case waterfall?

Reply:
The switch checks the `$this->current_token->tag` while this check looks into `$this->current_node()` – the most recent entry in the open elements stack. In other words, this closes any headings that are open when a heading tag opener is encountered.